File: 394-decode-string/decode-string.java
class Solution {
    public String decodeString(String s) {
        Stack<Integer> count=new Stack<>();
        Stack<StringBuilder> string=new Stack<>();
        StringBuilder curr=new StringBuilder();
        int num=0;

        for(char ch:s.toCharArray()){
            if(Character.isDigit(ch)){
                num=(num*10)+(ch-'0');
            }
            else if(ch=='['){
                count.push(num);
                string.push(curr);
                curr=new StringBuilder();
                num=0;
            }
            else if(ch==']'){
                int repeat=count.pop();
                StringBuilder prev=string.pop();
                for(int i=0;i<repeat;i++) prev.append(curr);
                curr=prev;
            }
            else curr.append(ch);
        }
        return curr.toString();
    }
}

File: 394-decode-string/decode-string.cpp
class Solution {
public:
    string decodeString(string s) {

        stack<string> st;
        stack<int> snum;
        string stemp="";
        string ans="";

        for (int i=0;i<s.size();i++){
            int num = 0;
            while (isdigit(s[i])){
                num = num*10 + (s[i]-'0');
                i++;
            }

            if (s[i]=='['){
                snum.push(num);
                st.push(stemp);
                stemp="";
            }
            else if (s[i]==']'){
                int n = snum.top();
                snum.pop();
                ans=st.top();
                st.pop();
                for (int i=0;i<n;i++){
                    ans = ans+ stemp ;
                }
                stemp=ans;
            }

            else stemp.push_back(s[i]);
        }
        return stemp;
    }
};